### DISCUSSION TITLE
Fatal exception during Engine.dispose()???

### DISCUSSION THREAD

**Discussion by MakoSDV** [sqlalchemy/sqlalchemy#10716] [Usage Questions] [answered]
I'm trying to migrate some code that worked in a Python 3.7 environment to work properly with an up-to-date Python 3.11 environment.

For some reason that I haven't been able to figure out, whenever an `sqlalchemy` `Engine` that had a connection is disposed of in the new environment, it causes Python to crash with a fatal exception code `0xc0000374`.

Originally, the exception was just happening during the garbage collection before the python process exits which didn't tell me much. However, I reproduced it with a more detailed stack trace by calling `Engine.dispose()`

for example, this basic code will result in the crash:
```python
from sqlalchemy import create_engine, text
sqlAlchemyURL = r"postgresql+psycopg2://user:pass@host/database"
dbengine = create_engine(sqlAlchemyURL)
with dbengine.connect() as conn:
    pass

dbengine.dispose()
```
I enabled the fault handler for python and was able to get these details about the exception:
```
Windows fatal exception: code 0xc0000374

Current thread 0x00001354 (most recent call first):
  File "C:\path\to\conda\env\Lib\site-packages\sqlalchemy\engine\default.py", line 701 in do_close
  File "C:\path\to\conda\env\Lib\site-packages\sqlalchemy\pool\base.py", line 379 in _close_connection
  File "C:\path\to\conda\env\Lib\site-packages\sqlalchemy\pool\base.py", line 885 in __close
  File "C:\path\to\conda\env\Lib\site-packages\sqlalchemy\pool\base.py", line 786 in close
  File "C:\path\to\conda\env\Lib\site-packages\sqlalchemy\pool\impl.py", line 215 in dispose
  File "C:\path\to\conda\env\Lib\site-packages\sqlalchemy\engine\base.py", line 3191 in dispose
  File "pMain.py", line 7 in <module>
```

Environment details:
Database: PostgreSQL 12.15

Python OS: Windows Server 2016

Updated Conda environment:
- python 3.11.6
- sqlalchemy 2.0.23
- psycopg2 2.9.7
- plus a bunch of other modules

The exception doesn't occur in the old Conda environment:
- python 3.7.16
- sqlalchemy 1.4.39
- psycopg2 2.9.3
- plus a bunch of other modules

I'm guessing it's some incompatibility with this combination of things in the setup, but it's not obvious what the actual problem is.
I'm still digging into it, so I'll update this if I learn more.

I was hoping that maybe someone in the community has some ideas on how I can more intelligently investigate it besides setting up a bunch of different environment and version permutations.

Thanks.

**Comment by CaselIT** [comment, marked as answer]
Hi,

You may also try to look into asking to the psycopg2 issue board, since this is likely to be something that comes from the driver.

Can you also reproduce using only the psycopg2 driver?

**Reply by MakoSDV** [reply]
Thanks.

I was able to reproduce it with both psycopg2 and the newer psycopg:

```python
import psycopg2
URL = r"postgresql://user:pass@host/database"
conn = psycopg2.connect(URL)
conn.close()
```

```
Windows fatal exception: code 0xc0000374

Current thread 0x000012e4 (most recent call first):
  File "pMain.py", line 4 in <module>
```

```python
import psycopg
URL = r"postgresql://user:pass@host/database"
conn = psycopg.connect(URL)
conn.close()
```

```
Windows fatal exception: code 0xc0000374

Current thread 0x00000c70 (most recent call first):
  File "C:\path\to\conda\env\Lib\site-packages\psycopg\connection.py", line 817 in close
  File "pMain.py", line 4 in <module>
```

I'll reach out to the psycopg community and see what they say. Thanks for your help.

**Comment by zzzeek** [comment]
IIUC Conda is doing the packaging here, I would assume they do the binary builds and all that so this would be on them.  this is an issue with the build of psycopg2
